### PR TITLE
Adding command panel for filtering

### PR DESCRIPTION
**Pull request by lawrencegripper**
This looks to allow users to input more complex commands via a panel similar to the command panel in VSCode. 

The initial use case is to filter the list view based on what is typed.

### CODE REVIEW

**Comment by lawrencegripper** [comment]
@stuartleeks have something ready... wasn't quite expecting the level of complexity involved. Want to sleep on things and see if I can do better tomorrow at refactoring some of this

filterazbrowse

Todo:
- Hook up to a keybinding to open the pallet 
- Hook up to a keybinding to launch pallet pre-populated for filtering ie with "/" at the start
- Work out kind with refresh 
- Currently requires two returns to exit panel should fix

File: internal/pkg/views/commandpanel.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	height := 2
+
+	v, err := g.SetView(w.name, w.x, w.y, w.x+w.w, w.y+height)
+	if err != nil && err != gocui.ErrUnknownView {
+		return err
+	}
+
+	v.Editable = true
+	g.Cursor = true
+	v.Title = "Command Pallet"
```
Comment (stuartleeks) [inline internal/pkg/views/commandpanel.go]:
```suggestion
	v.Title = "Command Palette"
```

File: internal/pkg/keybindings/keyhandlers.go [hunk 2]
```diff
@@ -6,32 +6,35 @@ import "github.com/jroimartin/gocui"
...
+	HandlerIDListPageDown            HandlerID = "listpagedown"        //nolint:golint
+	HandlerIDListPageUp              HandlerID = "listpageup"          //nolint:golint
+	HandlerIDListEnd                 HandlerID = "listend"             //nolint:golint
+	HandlerIDListHome                HandlerID = "listhome"            //nolint:golint
+	HandlerIDConfirmDelete           HandlerID = "confirmdelete"       //nolint:golint
+	HandlerIDClearPendingDeletes     HandlerID = "clearpendingdeletes" //nolint:golint
+	HandlerIDItemPageDown            HandlerID = "itempagedown"        //nolint:golint
+	HandlerIDItemPageUp              HandlerID = "itempageup"          //nolint:golint
+	HandlerIDToggleOpenCommandPanel  HandlerID = "opencommandpanel"    //nolint:golint
+	HandlerIDToggleCloseCommandPanel HandlerID = "closecommandpanel"   //nolint:golint
```
Comment (stuartleeks) [inline internal/pkg/keybindings/keyhandlers.go]:
```suggestion
	HandlerIDToggleCommandPanelOpen  HandlerID = "commandpanelopen"    //nolint:golint
	HandlerIDToggleCommandPanelClose HandlerID = "commandpanelclose"   //nolint:golint
```

Comment (stuartleeks) [reply internal/pkg/keybindings/keyhandlers.go]: Darn - multiline suggestions don'e work so well!

**Comment by jjcollinge** [comment]
Just trying this out - from a usage point of view I got a little confused. ESC/Ctrl + P toggles the panel, and Ctrl + F opens but doesn't hide. Once I've searched, if I press left or ENTER, it filters but then I seem to end up in a limbo where I can't navigate. What's the expected usage for this feature? Probably just being dumb!

**Comment by lawrencegripper** [comment]
@jjcollinge morning, sorry this is me being rubbish - I merged some unrelated changes and I think that's left things in a broken state. 

Flow I was aiming for was similar to VSCode. You could do `/` then `somefilter` and this would live filter the lefthand list then `enter` to close the `commandPanel` and `enter` again to open the top item on the list or `up/down` to move to one of the other items filtered.

File: internal/pkg/views/help.go [hunk 3]
```diff
@@ -2,83 +2,79 @@ package views
...
+
+	tmpl, err := template.New("help").Parse(tmplText)
+	if err != nil {
+		panic("Failed to parse help template. This is a Bug please raise an issue on GH. " + err.Error())
+	}
+
+	buf := new(bytes.Buffer)
+	err = tmpl.Execute(buf, keyBindings)
+	if err != nil {
+		panic("Failed to parse help template. This is a Bug please raise an issue on GH. " + err.Error())
```
Comment (stuartleeks) [inline internal/pkg/views/help.go]:
```suggestion
		panic("Failed to execute help template. This is a Bug please raise an issue on GH. " + err.Error())
```

**Comment by lawrencegripper** [comment]
Think this is nearly ready to go as a first attempt. What do we think?

File: internal/pkg/views/list.go [hunk 4]
```diff
@@ -33,8 +38,63 @@ type ListWidget struct {
...
+	listWidget := &ListWidget{ctx: ctx, x: x, y: y, w: w, h: h, contentView: contentView, statusView: status, enableTracing: enableTracing, lastTopIndex: 0, filterString: "", title: title, g: g}
+	go func() {
+		filterChannel := eventing.SubscribeToTopic("filter")
+		for {
+			filterStringInterface := <-filterChannel
+			filterString := strings.ToLower(strings.TrimSpace(strings.Replace(filterStringInterface.(string), "/", "", 1)))
+
+			filteredItems := []*handlers.TreeNode{}
+			for _, item := range listWidget.items {
+				if strings.Contains(strings.ToLower(item.Display), listWidget.filterString) {
```
Comment (stuartleeks) [inline internal/pkg/views/list.go]:
```suggestion
				if strings.Contains(strings.ToLower(item.Display), filterString) {
```

**Comment by stuartleeks** [comment]
LGTM

**Pull request merged**